File: api/src/cache/schemas.py
"""Pydantic request/response models for cache endpoints."""

from typing import Any

from pydantic import Field

from src.common.base_models import ApiModel

# ---------------------------------------------------------------------------
# Shared sub-models
# ---------------------------------------------------------------------------


class LookupConfig(ApiModel):
    """Configuration for a cache lookup request."""

    enable_exact_match: bool = True
    max_age_seconds: int | None = None


class CachedResponse(ApiModel):
    """The cached LLM response payload."""

    content: str
    model: str = ""
    tokens_used: dict[str, int] = Field(default_factory=dict)
    citations: list[dict[str, Any]] = Field(default_factory=list)


class WriteConfig(ApiModel):
    """Configuration for a cache write request."""

    ttl_seconds: int = 86400


class CacheMetadata(ApiModel):
    """Metadata about a cache entry returned on a hit."""

    created_at: str
    hit_count: int
    last_hit_at: str | None = None
    ttl_remaining_seconds: int | None = None


class LookupStages(ApiModel):
    """Timing breakdown of lookup pipeline stages."""

    exact_match_ms: float | None = None


# ---------------------------------------------------------------------------
# Lookup
# ---------------------------------------------------------------------------


class CacheLookupRequest(ApiModel):
    """POST /v1/cache/lookup request body."""

    workspace_id: str
    project_id: str
    query: str
    request_id: str | None = None
    lookup_config: LookupConfig = Field(default_factory=LookupConfig)


class CacheLookupResponse(ApiModel):
    """POST /v1/cache/lookup response body."""

    request_id: str | None = None
    status: str  # "hit" or "miss"
    source: str | None = None  # "exact" or "semantic"
    cache_entry_id: str | None = None
    response: CachedResponse | None = None
    similarity_score: float | None = None
    matched_query: str | None = None
    cache_metadata: CacheMetadata | None = None
    lookup_latency_ms: float = 0
    stages: LookupStages | None = None


# ---------------------------------------------------------------------------
# Write
# ---------------------------------------------------------------------------


class CacheWriteRequest(ApiModel):
    """POST /v1/cache/write request body."""

    workspace_id: str
    project_id: str
    query: str
    response: CachedResponse
    request_id: str | None = None
    write_config: WriteConfig = Field(default_factory=WriteConfig)


class CacheWriteResponse(ApiModel):
    """POST /v1/cache/write response body."""

    cache_entry_id: str
    request_id: str | None = None
    status: str = "written"
    stores: dict[str, str] = Field(default_factory=dict)
    expires_at: str | None = None
    created_at: str = ""


# ---------------------------------------------------------------------------
# Delete
# ---------------------------------------------------------------------------


class CacheDeleteResponse(ApiModel):
    """DELETE /v1/cache/entries/{id} response body."""

    cache_entry_id: str
    status: str = "invalidated"

File: ui/src/types/schemas.ts
import { z } from "zod";

// --- Shared ---

export const cachedResponseSchema = z.object({
  content: z.string(),
  model: z.string().default(""),
  tokensUsed: z.record(z.number()).default({}),
  citations: z.array(z.record(z.unknown())).default([]),
});

export const cacheMetadataSchema = z.object({
  createdAt: z.string(),
  hitCount: z.number(),
  lastHitAt: z.string().nullable().default(null),
  ttlRemainingSeconds: z.number().nullable().default(null),
});

export const lookupStagesSchema = z.object({
  exactMatchMs: z.number().nullable().default(null),
  embeddingMs: z.number().nullable().default(null),
  semanticMatchMs: z.number().nullable().default(null),
});

// --- Lookup ---

export const cacheLookupResponseSchema = z.object({
  requestId: z.string().nullable().default(null),
  status: z.string(),
  source: z.string().nullable().default(null),
  cacheEntryId: z.string().nullable().default(null),
  response: cachedResponseSchema.nullable().default(null),
  similarityScore: z.number().nullable().default(null),
  matchedQuery: z.string().nullable().default(null),
  cacheMetadata: cacheMetadataSchema.nullable().default(null),
  lookupLatencyMs: z.number().default(0),
  stages: lookupStagesSchema.nullable().default(null),
});

// --- Write ---

export const cacheWriteResponseSchema = z.object({
  cacheEntryId: z.string(),
  requestId: z.string().nullable().default(null),
  status: z.string().default("written"),
  stores: z.record(z.string()).default({}),
  expiresAt: z.string().nullable().default(null),
  createdAt: z.string().default(""),
});

// --- Invalidation ---

export const cacheInvalidateResponseSchema = z.object({
  requestId: z.string().nullable().default(null),
  entriesInvalidated: z.number(),
  invalidationCriteria: z.record(z.unknown()).default({}),
  createdAt: z.string().default(""),
});

// --- Purge ---

export const cachePurgeResponseSchema = z.object({
  requestId: z.string().nullable().default(null),
  entriesPurged: z.number(),
  scope: z.record(z.string()).default({}),
  createdAt: z.string().default(""),
});

// --- Config ---

export const cacheConfigSchema = z.object({
  enabled: z.boolean().default(true),
  defaultTtlSeconds: z.number().default(86400),
  semanticTtlSeconds: z.number().default(3600),
  similarityThreshold: z.number().default(0.92),
  maxEntrySizeBytes: z.number().default(102400),
  eventDrivenInvalidation: z.boolean().default(true),
  invalidationEvents: z.array(z.string()).default([]),
});

export const cacheConfigResponseSchema = z.object({
  workspaceId: z.string(),
  projectId: z.string(),
  config: cacheConfigSchema,
  updatedAt: z.string().default(""),
  updatedBy: z.string().nullable().default(null),
});

// --- Stats ---

export const tokensSavedSchema = z.object({
  input: z.number().default(0),
  output: z.number().default(0),
});

export const cacheStatsDetailSchema = z.object({
  totalLookups: z.number().default(0),
  exactHits: z.number().default(0),
  semanticHits: z.number().default(0),
  misses: z.number().default(0),
  hitRate: z.number().default(0),
  exactHitRate: z.number().default(0),
  semanticHitRate: z.number().default(0),
  totalEntries: z.number().default(0),
  estimatedCostSavedUsd: z.number().default(0),
  estimatedTokensSaved: tokensSavedSchema.default({}),
});

export const cacheStatsResponseSchema = z.object({
  workspaceId: z.string(),
  projectId: z.string(),
  period: z.string(),
  stats: cacheStatsDetailSchema,
});

// --- Delete ---

export const cacheDeleteResponseSchema = z.object({
  cacheEntryId: z.string(),
  status: z.string().default("invalidated"),
});

// --- Inferred types ---

export type CachedResponse = z.infer<typeof cachedResponseSchema>;
export type CacheMetadata = z.infer<typeof cacheMetadataSchema>;
export type CacheLookupResponse = z.infer<typeof cacheLookupResponseSchema>;
export type CacheWriteResponse = z.infer<typeof cacheWriteResponseSchema>;
export type CacheInvalidateResponse = z.infer<typeof cacheInvalidateResponseSchema>;
export type CachePurgeResponse = z.infer<typeof cachePurgeResponseSchema>;
export type CacheConfig = z.infer<typeof cacheConfigSchema>;
export type CacheConfigResponse = z.infer<typeof cacheConfigResponseSchema>;
export type CacheStatsDetail = z.infer<typeof cacheStatsDetailSchema>;
export type CacheStatsResponse = z.infer<typeof cacheStatsResponseSchema>;
export type CacheDeleteResponse = z.infer<typeof cacheDeleteResponseSchema>;
export type TokensSaved = z.infer<typeof tokensSavedSchema>;
